src/Cassandra/Metadata.cs

```diff
@@ -133,14 +148,120 @@ public IEnumerable<IPEndPoint> AllReplicas()
+                    return true;
+                }
+                Interlocked.Exchange(ref _lastRebuildTicks, currentTicks);
+                
+                Metadata.Logger.Info("Retrieving keyspaces metadata");
+                var ksMap = ksList.Select(ks => new KeyValuePair<string, KeyspaceMetadata>(ks.Name, ks));
+                _keyspaces = new ConcurrentDictionary<string, KeyspaceMetadata>(ksMap);
+                Metadata.Logger.Info("Rebuilding token map");
+                if (Partitioner == null)
+                {
+                    throw new DriverInternalError("Partitioner can not be null");
+                }
+                
+                _tokenMap = TokenMap.Build(Partitioner, Hosts.ToCollection(), _keyspaces.Values);
+                return true;
+            });
+
+            return await task.ConfigureAwait(false);
+        }
+
+        internal Task<bool> RemoveKeyspaceFromTokenMap(string name)
+        {
+            var currentTicks = DateTime.UtcNow.Ticks;
+            return _tokenMapTaskFactory.StartNew(
+                () =>
+                {
+                    var lastRebuildTicks = Interlocked.Read(ref _lastRebuildTicks);
+                    if (lastRebuildTicks > currentTicks)
+                    {
+                        return true;
+                    }
+                    
+                    if (_keyspacesLastUpdateTicks.TryGetValue(name, out var lastUpdateTicks) && lastUpdateTicks > currentTicks)
+                    {
+                        return true;
+                    }
+
+                    _keyspacesLastUpdateTicks[name] = currentTicks;
+                    
+                    Metadata.Logger.Verbose("Removing keyspace metadata: " + name);
+                    if (!_keyspaces.TryRemove(name, out var ks))
+                    {
+                        //The keyspace didn't exist
+                        return false;
+                    }
+                    _tokenMap?.RemoveKeyspace(ks);
+                    return true;
+                });
+        }
+
+        internal async Task<KeyspaceMetadata> UpdateTokenMapForKeyspace(string name)
         {
-            Logger.Info("Rebuilding token map");
-            if (Partitioner == null)
+            var currentTicks = DateTime.UtcNow.Ticks;
+            
+            // running this statement synchronously inside the exclusive scheduler deadlocks
+            var keyspaceMetadata = await _schemaParser.GetKeyspace(name).ConfigureAwait(false);
+
+            var task = _tokenMapTaskFactory.StartNew(
+                () =>
+                {
+                    var lastRebuildTicks = Interlocked.Read(ref _lastRebuildTicks);
+                    if (lastRebuildTicks > currentTicks)
+                    {
+                        return true;
+                    }
+                    
+                    if (_keyspacesLastUpdateTicks.TryGetValue(name, out var lastUpdateTicks) && lastUpdateTicks > currentTicks)
```

Comment:
I think this check can be made here, before fetching the metadata.

Reply:
It looks like I've sent the message without checking if this made sense, sorry... I'm the worst at multi tasking...

I was thinking about moving this check outside of the new task, but looking at it more closely I see that given that the operation is not atomic (read, compare value and then right), it must be done within the ordered task scheduler.